BoneCentral/Peripherals/Peripherals.SerialAdaptors/src/Serial.cpp

```diff
@@ -1,63 +1,191 @@
 //
 // Created by Nathan Copier on 1/28/2016.
+// Refactored by TekuConcept on 2/13/2017.
 //
 
 #include "Serial.h"
 
-Serial::Serial(std::shared_ptr<ScriptLogger> logger) : logger_(logger) {
-    std::system("stty -F /dev/ttyACM0 cs8 115200 ignbrk -brkint -icrnl -imaxbel -opost -onlcr -isig -icanon -iexten -echo -echoe -echok -echoctl -echoke noflsh -ixon -crtscts");
-    input_ = std::make_shared<std::ifstream>("/dev/ttyACM0");
-    output_ = std::make_shared<std::ofstream>("/dev/ttyACM0");
-    arduinoInitialized_ = false;
-    std::thread signalwait([](){ initializeArduino(); });
-}
+std::mutex Serial::serialLock_;
 
-void Serial::writeByte(unsigned short byteValue) {
-    if (arduinoInitialized_)
-    {
-        std::lock_guard<std::mutex> guard(serialLock_);
-        *output_ << char(byteValue & 0xFF) << std::flush;
+Serial::Serial(std::string device) {
+#ifdef DEBUG
+    LOG("\nReceived Device Name: " << device);
+    LOG("\nEntering Serial Debug Mode\n");
+    fd = 0;
+#else
+    if((fd = open(device.c_str(), O_RDWR)) < 0) {
+        LOG("Device failed to open... entering dummy mode.\n");
+        fd = 0;
+        return;
     }
+    
+    configure();
+#endif
+    ackno();
 }
 
-void Serial::writeShort(unsigned short shortValue) {
-    if (arduinoInitialized_)
-    {
-        std::lock_guard<std::mutex> guard(serialLock_);
-        *output_ << char(shortValue >> 8) << char(shortValue & 0xFF) << std::flush;
-    }
+Serial::~Serial() {
+    if(fd != 0) close(fd);
 }
 
-json Serial::readJson() {
-    if (arduinoInitialized_)
-    {
-        std::lock_guard<std::mutex> guard(serialLock_);
-        json arduinoData;
-        *input >> arduinoData;
-        return arduinoData;
+void Serial::configure() {
+    struct termios topts;
+    if(tcgetattr(fd, &topts)) {
+        LOG("Failed to get terminal ios options from file descriptor.\n");
+        throw 1;
+    }
+    if(cfsetispeed(&topts, B115200)) {
+        LOG("Failed to set input baud rate.\n");
+        throw 1;
+    }
+    if(cfsetospeed(&topts, B115200)) {
+        LOG("Failed to set output baud rate.\n");
+        throw 1;
+    }
+    topts.c_cflag &= ~(PARENB|CSTOPB|CSIZE|CRTSCTS);
+    topts.c_cflag |= CS8|CREAD|CLOCAL;
+    topts.c_iflag &= ~(IXON | IXOFF | IXANY);
+    topts.c_lflag &= ~(ICANON | ECHO | ECHOE | ISIG);
+    topts.c_oflag &= ~OPOST;
+    topts.c_cc[VMIN] = 1;
+    topts.c_cc[VTIME] = 0;
+    if(tcsetattr(fd, TCSANOW, &topts)) {
+        LOG("Failed to set terminal ios options for file descriptor.\n");
+        throw 1;
+    }
+    if(tcflush(fd, TCIFLUSH)) {
+        LOG("Failed to flush file descriptor.\n");
+        throw 1;
```

Comment:
To be honest, this was a copy-past from Daniel Tonk's code. Other than the `cfsetispeed` and `cfsetospeed`, I'm not exactly sure how else I can break it up. All the c_Xflag (configuration flags) options are the same options passed in the original system call.